### PR TITLE
Add support for return statements in conditionals

### PR DESCRIPTION
**Pull request by dime10**
[sc-57417]

### CODE REVIEW

File: doc/changelog.md [hunk 1]
```diff
@@ -4,6 +4,30 @@
 
 <h3>Improvements</h3>
 
+* Catalyst now supports return statements inside conditionals in `@qjit(autograph=True)` compiled
+  functions.
+  [(#)]()
```
Comment (josh146) [inline doc/changelog.md:9]:
```suggestion
  [(#583)](https://github.com/PennyLaneAI/catalyst/pull/583)
```

File: doc/changelog.md [hunk 2]
```diff
@@ -4,6 +4,30 @@
...
+
+  ```py
+  @qjit(autograph=True)
+  @qml.qnode(qml.device(backend, wires=1))
+  def f(x: float):
+    qml.RY(x, wires=0)
+
+    m = measure(0)
+    if not m:
+        return 0
```
Comment (josh146) [inline doc/changelog.md:21]: 💯

File: doc/changelog.md [hunk 3]
```diff
@@ -4,6 +4,30 @@
...
 <h3>Improvements</h3>
 
+* Catalyst now supports return statements inside conditionals in `@qjit(autograph=True)` compiled
+  functions.
+  [(#)]()
+
+  The following is now possible:
+
+  ```py
+  @qjit(autograph=True)
```
Comment (josh146) [inline doc/changelog.md:14]: I wonder why the syntax highlighting is not working 🤔

File: doc/changelog.md [hunk 4]
```diff
@@ -4,6 +4,30 @@
...
+  @qml.qnode(qml.device(backend, wires=1))
+  def f(x: float):
+    qml.RY(x, wires=0)
+
+    m = measure(0)
+    if not m:
+        return 0
+
+    ...
+
```
Comment (josh146) [inline doc/changelog.md:24]:
```suggestion
```

Comment (josh146) [reply doc/changelog.md:24]: Just to make it an example that copies and pastes without error

File: doc/dev/autograph.rst [hunk 5]
```diff
@@ -173,40 +173,45 @@ converted, there are some important constraints and restrictions to be aware of.
...
-    def f(x):
-        if x > 5:
-            y = x ** 2
-        else:
-            y = x ** 3
-        return y
+>>> qjit(autograph=True)(f)
+TypeError: Conditional requires a consistent return structure across all branches!
+Got PyTreeDef((*, CustomNode(ExpectationMP[(('wires', None),)], [CustomNode(PauliZ[(<Wires = [0]>, ())], []), None])))
+and PyTreeDef((*, CustomNode(ExpectationMP[(('wires', None),)], [CustomNode(PauliY[(<Wires = [0]>, ())], []), None]))).
```
Comment (josh146) [inline doc/dev/autograph.rst:214]: thanks for updating this!

**Review by josh146 — APPROVED** [review]
Nice, thanks @dime10!

File: doc/changelog.md (hunk 4 above)
Comment (dime10) [reply doc/changelog.md:24]: `...` is a valid Python expression :D

File: doc/dev/autograph.rst [hunk 6]
```diff
@@ -173,40 +173,45 @@ converted, there are some important constraints and restrictions to be aware of.
...
-This will not be correctly captured by AutoGraph, and instead will be
-interpreted as
+This will generate the following error:
+
+>>> qjit(autograph=True)(f)
+TypeError: Conditional requires a consistent array shape per result across all branches!
+Got () for result #1 but expected (2,).
+
+Another example is the use of different *structure* across branches. The structure of function
+result is defined by things like the number of results, the containers used like lists or
```
Comment (dime10) [inline doc/dev/autograph.rst:196]:
```suggestion
Another example is the use of different *structure* across branches. The structure of a function
output is defined by things like the number of results, the containers used like lists or
```

File: doc/dev/autograph.rst [hunk 7]
```diff
@@ -173,40 +173,45 @@ converted, there are some important constraints and restrictions to be aware of.
...
+dictionaries, or more generally any (compile-time) PyTree metadata. For PennyLane, this means
+returning different observables for example is also not supported, as the observable class is
```
Comment (dime10) [inline doc/dev/autograph.rst:198]:
```suggestion
returning different observables for example is not supported, as the observable class is
```

File: doc/changelog.md (hunk 3 above)
Comment (dime10) [reply doc/changelog.md:14]: On GitHub I'm not sure, it doesn't seem to syntax highlight code blocks in the code view of markdown files. It is highlight in the rendered version on GitHub or the docs though.

File: doc/changelog.md [hunk 8]
```diff
@@ -4,6 +4,30 @@
...
 
+* Catalyst now supports return statements inside conditionals in `@qjit(autograph=True)` compiled
+  functions.
+  [(#)]()
+
+  The following is now possible:
+
+  ```py
+  @qjit(autograph=True)
+  @qml.qnode(qml.device(backend, wires=1))
```
Comment (dime10) [inline doc/changelog.md:15]:
```suggestion
  @qml.qnode(qml.device("lightning.qubit", wires=1))
```

Comment (dime10) [inline doc/changelog.md:15]:
```suggestion
  @qml.qnode(qml.device("lightning.qubit", wires=1))
```

File: doc/changelog.md (hunk 2 above)
Comment (dime10) [inline doc/changelog.md:21]: Almost contradicted the requirements in my example 😅 Unfortunate that you can't (easily) sub in constants for a measurement process though:
```suggestion
    if not m:
        return qml.expval(qml.PauliZ(0))
```

**Pull request merged**